### PR TITLE
User authentication

### PR DESCRIPTION
**Pull request by cloudbeatsch**
Changed the userId to be sent from trusted services

### CODE REVIEW

File: README.MD [hunk 1]
```diff
@@ -10,6 +10,7 @@ Fill in the following values:
 
 `CONTRACT_ADDRESS`: Your contract address, see details below
 `ACCOUNT_ADDRESS` : Your account address, see details below
+`USE_TESTRPC` : true if testRpc is used, otherwise false
```
Comment (amiturgman) [inline README.MD]: We don't need the USE_TESTRPC in the configuration, this is a native environment variable.

If you want to know if you're using testRpc, can use the utils.js file:
var utils = require('./utils');
utils.isTestRpc

I don't see that you use this anyway in the rest of the code, so this is probably leftovers from the previous PR.

File: services/proof.js [hunk 2]
```diff
@@ -7,17 +7,19 @@ var key = require('./key');
 var contract = require('./contract');
 var config = require('../config');
 
-// TODO move userId to options object in each API
-const userId = "un-authenticated";
-
-
+function makeValidUserId(userId){
```
Comment (amiturgman) [inline services/proof.js:10]: how does the userId look like that we need this check?

**Comment by amiturgman** [comment]
@cloudbeatsch 
The problem with the solution as it is now, is that anyone who has anyone else's user Id, can send request on his behalf, since you extract the user id in the office part, and call ibera-services api with the id in the query string.

I guess that what we want to do is to add a few authentication schemes to the ibera-services api (office365, google, etc..), and send the user's token in the headers to the services api. only there we should extract the user Id internally before processing the request. The way you implemented it now is not secure, unless you we use client-server certificates, but then you have other issues that you'll have to install the client certificate in each client and I don't think this is what you intended...

Let me know what you think...

**Comment by cloudbeatsch** [comment]
@amiturgman agree on the possible attack surface for userId injection. I was thinking about adding a token for the service but decided against it to keep it simple. The constraint is that only trusted services have access to the ibera-services - this must be enforced through the VPNs network security rules. As is, only the api part (server side node part) of the office  addIn can access ibera-services. The client part of the office-integration doesn't send the userId but the whole user-token as a header to the document-service. For the Android client, the idea would be to use a token between the client and the gatway, where the gateway is a trusted caller.

File: services/proof.js (hunk 2 above)
Comment (cloudbeatsch) [reply services/proof.js:10]: currently it's a guid but it might be anything that Exchange returns

File: README.MD (hunk 1 above)
Comment (cloudbeatsch) [reply README.MD]: yep that's legacy

**Comment by amiturgman** [comment]
@cloudbeatsch so there's a missing part which is the android gateway that we didn't think/talk about... cause I thought that the android device was supposed to talk directly to the services api... (and so any other client).
This means we need to add a backend service to the android client that will be deployed as part of the solution in the App Service Environment. And will be able to communicate with the services api internally. We should also remove the public facing endpoint from the services API.
I don't know... it sounds too complicated...
Usually in such cases, the main api that everyone talks to (the services api in our case) supports a few authentication schemes, and based on the token in the header, use the correct one to resolve the user's identity. This can be done with the passport node module (you can see an example of how I did it with google auth [here](https://github.com/amiturgman/web-cli-sample-app/blob/master/app.js#L22)).
We can decide to leave it as is, for the sake of simplicity and saying that this is a POC. But as an E2E solution, the authentication must be addressed... I don't mind also leaving it as you suggested, it just that it makes things much more complicated technically-wize.

**Comment by cloudbeatsch** [comment]
If you prefer to connect from the Android directly to the service, we should definitely move towards token based authentication  - let's add this as an issue and address once we move off the POC, ok?

**Comment by amiturgman** [comment]
@cloudbeatsch sure, thanks.

**Pull request merged**